### PR TITLE
 EEI: implement getReturnDataSize and returnDataCopy

### PR DESCRIPTION
**Pull request by axic**

### CODE REVIEW

File: src/eei.cpp [hunk 1]
```diff
@@ -468,9 +477,9 @@ namespace HeraVM {
       context->fn_table->call(&call_result, context, &call_message);
 
       if (call_result.output_data) {
-        vector<uint8_t> result(call_result.output_data, call_result.output_data + call_result.output_size);
-        result.resize(resultLength);
-        storeMemory(result, 0, resultOffset, resultLength);
+        lastReturnData.assign(call_result.output_data, call_result.output_data + call_result.output_size);
```
Comment (jakelang) [inline src/eei.cpp]: Is call not supposed to store output data at resultOffset anyway?

Comment (axic) [reply src/eei.cpp]: There is no point doing that with returndata anymore. See https://github.com/ewasm/design/issues/70

Comment (axic) [reply src/eei.cpp]: But you are correct, perhaps I should separate these changes into two: supporting returndata and removing that option.

Comment (axic) [reply src/eei.cpp]: Done, that will be a different PR.

File: src/eei.cpp [hunk 2]
```diff
@@ -401,6 +401,23 @@ namespace HeraVM {
       return Literal();
     }
 
+    if (import->base == Name("returnDataSize")) {
+      HERA_DEBUG << "returnDataSize\n";
+      return Literal((uint32_t)lastReturnData.size());
```
Comment (chfast) [inline src/eei.cpp]: Use `static_cast`.

File: src/eei.cpp [hunk 3]
```diff
@@ -475,6 +492,10 @@ namespace HeraVM {
         vector<uint8_t> result(call_result.output_data, call_result.output_data + call_result.output_size);
         result.resize(resultLength);
         storeMemory(result, 0, resultOffset, resultLength);
+
+        lastReturnData.assign(call_result.output_data, call_result.output_data + call_result.output_size);
+      } else {
+        lastReturnData.resize(0);
```
Comment (chfast) [inline src/eei.cpp]: 1. `.resize(0)` === `.clear()`.
2. You have to clear it unconditionally.

File: src/eei.h [hunk 4]
```diff
@@ -104,6 +104,7 @@ struct EthereumInterface : ShellExternalInterface {
   struct evm_context* context = nullptr;
   std::vector<uint8_t> const& code;
   struct evm_message const& msg;
+  std::vector<uint8_t> lastReturnData = std::vector<uint8_t>();
```
Comment (chfast) [inline src/eei.h]: You need a default initializer only for "primitive" types.

File: src/eei.cpp [hunk 5]
```diff
@@ -475,6 +492,10 @@ namespace HeraVM {
         vector<uint8_t> result(call_result.output_data, call_result.output_data + call_result.output_size);
         result.resize(resultLength);
         storeMemory(result, 0, resultOffset, resultLength);
+
+        lastReturnData.assign(call_result.output_data, call_result.output_data + call_result.output_size);
```
Comment (chfast) [inline src/eei.cpp]: The same goes for CREATE.

File: src/eei.cpp [hunk 6]
```diff
@@ -401,6 +401,23 @@ namespace HeraVM {
       return Literal();
     }
 
+    if (import->base == Name("returnDataSize")) {
```
Comment (axic) [inline src/eei.cpp]: Should be `getReturnDataSize`.

**Pull request merged**